File: distribute-coins-in-binary-tree/distribute-coins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        def balance(node):
            nonlocal res
            if not node: return 0
            left = balance(node.left)
            right = balance(node.right)
            res += abs(left) + abs(right)
            
            return node.val - 1 + left + right
        
        res = 0
        balance(root)
        return res 
    
        

File: distribute-coins-in-binary-tree/distribute-coins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int distributeCoins(TreeNode* root) {
        balance(root, res);
        return res;
    }
private:
    int res = 0;
    int balance(TreeNode* node, int& res){
        if (!node) return 0;
        int left = balance(node->left, res);
        int right = balance(node->right, res);
        res += abs(left) + abs(right);
        return node->val - 1 + left + right;
    }
};
